### PR TITLE
Add Quantum GANs Demo

### PR DESCRIPTION
**Pull request by jamesellis1999**
Title: Quantum GANs for Image Generation

Summary: A demonstration of a Quantum GAN method recently proposed in the literature (see relevant references). We use the method to generate hand-written digits of zero. 

Relevant references: https://arxiv.org/pdf/2010.06201.pdf

Possible Drawbacks: None

Related GitHub Issues: None

### CODE REVIEW

File: demos_qml.rst [hunk 1]
```diff
@@ -91,6 +91,12 @@ IBM, Google, Rigetti, and many more—or on real hardware devices.
     :figure: demonstrations/ensemble_multi_qpu/ensemble_diagram.png
     :description: :doc:`demos/tutorial_ensemble_multi_qpu`
```
Comment (josh146) [inline demos_qml.rst:92]:
```suggestion
    :description: :doc:`demos/ensemble_multi_qpu`
```

I noticed a broken link here :)

File: demos_qml.rst [hunk 2]
```diff
@@ -91,6 +91,12 @@ IBM, Google, Rigetti, and many more—or on real hardware devices.
...
     :tags: pytorch forest qiskit
+    
+.. customgalleryitem::
+    :tooltip: Generate images with Quantums GANs.
+    :figure: demonstrations/quantum_gans/patch.jpeg
+    :description: :doc:`demos/quantum_gans`
```
Comment (josh146) [inline demos_qml.rst:98]:
```suggestion
    :description: :doc:`demos/tutorial_quantum_gans`
```
this needs to match the name of the file containing the tutorial :)

**Comment by josh146** [comment]
Thank you for this wonderful contribution @jamesellis1999! Someone from the team will be on hand to review it shortly.

In the meantime, you can click the 'Details' button next door to the `build artifact` CI when it is complete, to browse the built version of your demo :)

image

**Comment by jamesellis1999** [comment]
Thanks @josh146! Are your changes above already implemented, or do I have to change them myself? Sorry, I am new to this. 

Secondly, when on the mobile version of my tutorial, all images and some equations overflow the main page (see attached). Is there a way to fix this?

image

image

**Comment by josh146** [comment]
@jamesellis1999 I think this can be fixed by using `:width: 90%` (or whatever width value looks best for your images), rather than `:scale:`

File: demonstrations/tutorial_quantum_gans.py [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+
+
+######################################################################
+# Generative Adversarial Networks (GANs)
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+
+
+######################################################################
+# The goal of generative adversarial networks (GANs) [1] is to generate
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:31]: This is a relatively minor comment :) However, if you write the references section at the bottom with _named_ labels, for example,

https://github.com/PennyLaneAI/qml/blob/c14e495265598cc927f29e89a9597c168bc52958/demonstrations/tutorial_expressivity_fourier_series.py#L822-L830

Then you can cite the reference in text as so,

https://github.com/PennyLaneAI/qml/blob/c14e495265598cc927f29e89a9597c168bc52958/demonstrations/tutorial_expressivity_fourier_series.py#L631

and the website will automatically generate a hyperlink.

File: demonstrations/tutorial_quantum_gans.py [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+# -  :math:`D(\boldsymbol{x})`: probability of the discriminator
+#    classifying real data as real
+# -  :math:`G(\boldsymbol{z})`: fake data
+# -  :math:`D(G(\boldsymbol{z}))`: probability of discriminator
+#    classifying fake data as real
+#
+# In practice, the two networks are trained iteratively, each with
+# a separate loss function to be minimised,
+#
+# .. math::  L_D = -[y \cdot log(D(x)) + (1-y)\cdot log(1-D(G(z)))]
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:73]:
```suggestion
# .. math::  L_D = -[y \cdot \log(D(x)) + (1-y)\cdot \log(1-D(G(z)))]
```

File: demonstrations/tutorial_quantum_gans.py [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+#
+# .. math::  L_G = [(1-y) \cdot log(1-D(G(z)))]
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:75]:
```suggestion
# .. math::  L_G = [(1-y) \cdot \log(1-D(G(z)))]
```

File: demonstrations/tutorial_quantum_gans.py [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+#
+# where :math:`y` is a binary label for real (:math:`y=1`) or fake
+# (:math:`y=0`) data. In practice, generator training is shown to be more
+# stable [1] when made to maximise :math:`log(D(G(z)))` instead of
+# minimising :math:`log(1-D(G(z)))`. Hence, the generator loss function to
+# be minimised becomes,
+#
+# .. math::  L_G = -[(1-y) \cdot log(D(G(z)))]
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:83]:
```suggestion
# .. math::  L_G = -[(1-y) \cdot \log(D(G(z)))]
```

File: demonstrations/tutorial_quantum_gans.py [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+# a separate loss function to be minimised,
+#
+# .. math::  L_D = -[y \cdot log(D(x)) + (1-y)\cdot log(1-D(G(z)))]
+#
+# .. math::  L_G = [(1-y) \cdot log(1-D(G(z)))]
+#
+# where :math:`y` is a binary label for real (:math:`y=1`) or fake
+# (:math:`y=0`) data. In practice, generator training is shown to be more
+# stable [1] when made to maximise :math:`log(D(G(z)))` instead of
+# minimising :math:`log(1-D(G(z)))`. Hence, the generator loss function to
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:80]:
```suggestion
# stable [1] when made to maximise :math:`\log(D(G(z)))` instead of
# minimising :math:`\log(1-D(G(z)))`. Hence, the generator loss function to
```

File: demonstrations/tutorial_quantum_gans.py [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+# In this tutorial, we re-create one of the quantum GAN methods presented
+# by Huang et al.[2]: the patch method. This method uses several quantum
+# generators, with each sub-generator, :math:`G^{(i)}`, responsible for
+# constructing a small patch of the final image. The final image is
+# contructed by concatenting all of the patches together as shown below.
+#
+# .. figure:: ../demonstrations/quantum_gans/patch.jpeg
+#   :width: 90% 
+#   :alt: quantum_patch_method
+#   :align: center
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:103]: This is a great image! really helps to explain the patch method.

File: demonstrations/tutorial_quantum_gans.py [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+
+transform = transforms.Compose([transforms.ToTensor()])
+dataset = DigitsDataset(csv_file="quantum_gans/optdigits.tra", transform=transform)
+dataloader = torch.utils.data.DataLoader(
+    dataset, batch_size=batch_size, shuffle=True, drop_last=True
+)
+
+
+######################################################################
+# Let's visual some data.
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:215]:
```suggestion
# Let's visualize some of the data.
```

File: demonstrations/tutorial_quantum_gans.py [hunk 10]
```diff
@@ -0,0 +1,595 @@
...
+
+for i in range(8):
+    image = dataset[i][0].reshape(image_size,image_size)
+    plt.subplot(1,8,i+1)
+    plt.axis('off')
+    plt.imshow(image.numpy(), cmap='gray')
+    
+plt.show()
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:223]: I wonder if `plt.tight_layout()` used before the `show` will help remove some of the whitespace?

File: demonstrations/tutorial_quantum_gans.py [hunk 11]
```diff
@@ -0,0 +1,595 @@
...
+# process from latent vector input to image output can be split into four
+# distinct sections: state embedding, parameterisation, non-linear
+# transformation, and post-processing. Each of the following sections
+# below refer to a single iteration of the training process to simplify
+# the discussion.
+#
+# .. figure:: ../demonstrations/quantum_gans/qcircuit.jpeg
+#   :width: 90% 
+#   :alt: quantum_circuit
+#   :align: center
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:280]: Another great image!

File: demonstrations/tutorial_quantum_gans.py [hunk 12]
```diff
@@ -0,0 +1,595 @@
...
+#
+# **1.) State Embedding**
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:282]: I suggest removing either the `.` or the `)` here and below, and just going with one delimiter.

File: demonstrations/tutorial_quantum_gans.py [hunk 13]
```diff
@@ -0,0 +1,595 @@
...
+# by,
+#
+# .. math:: |\Psi(z)\rangle = U_{G}(\theta)|\boldsymbol{z}\rangle
+#
+# where :math:`U_{G}(\theta)` represents the overall unitary of the
+# parameterised layers. Let us inspect the state when we take a partial
+# measurment, :math:`\Pi`, and trace out the ancillary subsystem,
+# :math:`\mathcal{A}`,
+#
+# .. math:: \rho(\boldsymbol{z}) = \frac{Tr_{\mathcal{A}}(\Pi \otimes \mathbb{I} |\Psi(z)\rangle \langle \Psi(\boldsymbol{z})|) }{Tr(\Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle \langle \Psi(\boldsymbol{z})|))} = \frac{Tr_{\mathcal{A}}(\Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle \langle \Psi(\boldsymbol{z})|) }{\langle \Psi(\boldsymbol{z})| \Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle}
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:315]:
```suggestion
# .. math:: \rho(\boldsymbol{z}) = \frac{\text{Tr}_{\mathcal{A}}(\Pi \otimes \mathbb{I} |\Psi(z)\rangle \langle \Psi(\boldsymbol{z})|) }{\text{Tr}(\Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle \langle \Psi(\boldsymbol{z})|))} = \frac{\text{Tr}_{\mathcal{A}}(\Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle \langle \Psi(\boldsymbol{z})|) }{\langle \Psi(\boldsymbol{z})| \Pi \otimes \mathbb{I} |\Psi(\boldsymbol{z})\rangle}
```

File: demonstrations/tutorial_quantum_gans.py [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+# Due to the normalisation constraint of the measurment, all elements in 
+# :math:`\boldsymbol{g}^{(i)}` must sum to one. This is
+# a problem if we are to use :math:`\boldsymbol{g}^{(i)}` as the pixel
+# intensity values for our patch. For example, imagine a hypothetical
+# situation where a patch of full intensity pixels was the target. The
+# best patch a sub-generator could produce would be a patch of pixels all
+# at a magnitude of :math:`\frac{1}{2^{N-N_A}}`. To alleviate this
+# constraint, we apply a post-processing technique to each patch,
+#
+# .. math::  \boldsymbol{\tilde{x}^{(i)}} = \frac{\boldsymbol{g}^{(i)}}{max_{k}\boldsymbol{g}_k^{(i)}}
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:342]:
```suggestion
# .. math::  \boldsymbol{\tilde{x}^{(i)}} = \frac{\boldsymbol{g}^{(i)}}{\max_{k}\boldsymbol{g}_k^{(i)}}
```

File: demonstrations/tutorial_quantum_gans.py [hunk 15]
```diff
@@ -0,0 +1,595 @@
...
+
+######################################################################
+#.. note::
+#
+#    You may have noticed ``errG = criterion(outD_fake, real_labels)`` and
+#    wondered why we don’t use ``fake_labels`` instead of ``real_labels``.
+#    However, this is simply a trick to be able to use the same
+#    ``criterion`` function for both the generator and discriminator.
+#    Using ``real_labels`` forces the generator loss function to use the
+#    :math:`log(D(G(z))` term instead of the :math:`log(1 - D(G(z))` term
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:541]:
```suggestion
#    :math:`\log(D(G(z))` term instead of the :math:`\log(1 - D(G(z))` term
```

File: demonstrations/tutorial_quantum_gans.py [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+
+        ax = plt.Subplot(fig, inner[j])
+        ax.imshow(im.numpy(), cmap="gray")
+        ax.set_xticks([])
+        ax.set_yticks([])
+        if j==0:
+            ax.set_title(f'Iteration {50+i*50}', loc='left')
+        fig.add_subplot(ax)
+
+plt.show()
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:566]: Nice, it looks like it trained quite well!

File: demonstrations/tutorial_quantum_gans.py [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+# ~~~~~~~~~~
+#
+
+
+######################################################################
+# [1] Ian J. Goodfellow et al. *Generative Adversarial Networks*.
+# arXiv:1406.2661 (2014).
+#
+# [2] He-Liang Huang et al. *Experimental Quantum Generative Adversarial
+# Networks for Image Generation*. arXiv:2010.06201 (2020).
```
Comment (josh146) [inline demonstrations/tutorial_quantum_gans.py:594]: Relatively minor, but could be good to make the arXiv identifiers links to the arXiv papers

**Review by josh146 — COMMENTED** [review]
Thanks @jamesellis1999! A great tutorial 💪 I've gone through and left a few comments and suggestions

File: demonstrations/tutorial_quantum_gans.py (hunk 3 above)
Comment (jamesellis1999) [reply demonstrations/tutorial_quantum_gans.py:38]: I have noticed this method gives some weird formatting in the rendered webpage. If the reference is mentioned 5 times in the main text, you get (1,2,3,4,5) in front of the reference (e.g references sections of https://pennylane.ai/qml/demos/tutorial_expressivity_fourier_series.html#schuld2020).  Is there anyway to prevent this?

Comment (josh146) [reply demonstrations/tutorial_quantum_gans.py:38]: Ah, this is a 'feature' in a sense, that Sphinx is linking to all 5 citations separately (which seems odd on face value if you are used to latex-style citation crossreferences).

Unfortunately, there is no way to disable this --- feel free to keep the current version if you prefer it

Comment (jamesellis1999) [reply demonstrations/tutorial_quantum_gans.py:38]: I think I will keep the current referencing style if that is ok. I quite like the latex-style.

**Comment by jamesellis1999** [comment]
That should be all of the suggestions fixed! Thanks @josh146, let me know if there is anything else I could fix.

**Review by josh146 — APPROVED** [review]
Hi @jamesellis1999, it looks great from my side. Thank you for this contribution! 💪
